File: clients/node/src/errno.ts
// Platform-specific errno constants for parity with the Rust
// `varta-client::classify_send_error`. Mirrors Python's `_errno.py`.
//
// Node exposes runtime errno values via `os.constants.errno`, but the
// `ENOBUFS` symbol is not present on every platform's constant table
// (Linux=105, BSD/macOS=55, illumos/Solaris=132). We hard-code per
// `process.platform` to match the Rust constants exactly.

import { constants as osConstants } from "node:os";

const LINUX_ENOBUFS = 105;
const BSD_ENOBUFS = 55; // macOS / iOS / FreeBSD / NetBSD / OpenBSD / DragonFly
const SOLARIS_ENOBUFS = 132;

function selectEnobufs(): number {
  const p = process.platform;
  if (p === "linux") return LINUX_ENOBUFS;
  if (p === "darwin" || p === "freebsd" || p === "openbsd") {
    return BSD_ENOBUFS;
  }
  if (p === "sunos") return SOLARIS_ENOBUFS;
  // Fall back to whatever the running platform exposes — better than
  // silently using the Linux value on an unknown OS.
  const fromOs = osConstants.errno as Record<string, number>;
  return fromOs.ENOBUFS ?? LINUX_ENOBUFS;
}

const errno = osConstants.errno as Record<string, number>;

export const ENOBUFS = selectEnobufs();
export const ENOSPC = errno.ENOSPC;
export const EAGAIN = errno.EAGAIN;
export const EWOULDBLOCK = errno.EWOULDBLOCK ?? errno.EAGAIN;
export const ECONNREFUSED = errno.ECONNREFUSED;
export const ECONNRESET = errno.ECONNRESET;
export const ENOTCONN = errno.ENOTCONN;
export const EPIPE = errno.EPIPE;
export const ENOENT = errno.ENOENT;

// Reverse-lookup table for symbolic errno names. Used when surfacing
// unexpected `BeatOutcome.failed` results so the caller's log line has
// "ENOENT" instead of just "errno=2".
const NAME_BY_CODE: Map<number, string> = (() => {
  const m = new Map<number, string>();
  for (const [name, code] of Object.entries(errno)) {
    if (!m.has(code as number)) m.set(code as number, name);
  }
  return m;
})();

export function errnoName(code: number | null | undefined): string {
  if (code === null || code === undefined || code === 0) return "Unknown";
  return NAME_BY_CODE.get(code) ?? `errno_${code}`;
}

File: clients/python/src/varta/_errno.py
"""Platform-specific errno constants required for parity with the Rust
``varta-client::classify_send_error``.

These values are hard-coded so the base ``pip install varta`` install
remains stdlib-only on every supported platform. The Rust client makes
the same choice (`crates/varta-client/src/client.rs:22-60`) and gates
on compile-time targets. Python sees ``platform.system()`` at runtime.

If you are porting the client to a new platform, add the ``ENOBUFS``
value here and extend :data:`SUPPORTED_PLATFORMS`.
"""

from __future__ import annotations

import errno as _errno
import sys

__all__ = [
    "ENOBUFS",
    "ENOSPC",
    "EAGAIN",
    "EWOULDBLOCK",
    "ECONNREFUSED",
    "ECONNRESET",
    "ENOTCONN",
    "EPIPE",
    "ENOENT",
    "UnsupportedPlatformError",
]


class UnsupportedPlatformError(RuntimeError):
    """Raised at import time when running on an unsupported platform."""


_LINUX_ENOBUFS = 105
_BSD_ENOBUFS = 55  # macOS / iOS / FreeBSD / NetBSD / OpenBSD / DragonFly
# Solaris / illumos ENOBUFS = 132 (confirmed against rust-libc
# ``src/unix/solarish/mod.rs``; matches the Rust client and the Node client).
# NOT 111 — that is Linux's ECONNREFUSED and is undefined on solarish, so the
# old value silently routed real send-buffer backpressure to ``failed`` instead
# of ``Dropped(KERNEL_QUEUE_FULL)`` (this is the Python sibling of Rust bug-470).
_SOLARIS_ENOBUFS = 132  # Solaris / illumos


def _select_enobufs() -> int:
    platform = sys.platform
    if platform.startswith("linux"):
        return _LINUX_ENOBUFS
    if platform in ("darwin", "ios") or platform.startswith(
        ("freebsd", "netbsd", "openbsd", "dragonfly")
    ):
        return _BSD_ENOBUFS
    if platform in ("sunos5", "solaris", "illumos"):
        return _SOLARIS_ENOBUFS
    # Fall back to the Python `errno` module if it knows the symbol on
    # this platform — gives us a chance on platforms we did not list
    # explicitly without silently using the wrong number.
    return getattr(_errno, "ENOBUFS", _LINUX_ENOBUFS)


ENOBUFS: int = _select_enobufs()
ENOSPC: int = _errno.ENOSPC
EAGAIN: int = _errno.EAGAIN
EWOULDBLOCK: int = _errno.EWOULDBLOCK
ECONNREFUSED: int = _errno.ECONNREFUSED
ECONNRESET: int = _errno.ECONNRESET
ENOTCONN: int = _errno.ENOTCONN
EPIPE: int = _errno.EPIPE
ENOENT: int = _errno.ENOENT
